include/macros.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+const int ROWS = 8;
+const int COLS = 8;
+
+struct Dir
+{
+	int x;
+	int y;
+};
+
+struct Location {
+	int x;
+	int y;
+
+	bool operator==(const Location& other) const {
+		return x == other.x && y == other.y;
+	}
+
+	bool operator!=(const Location& other) const {
+		return x != other.x || y != other.y;
+	}
+
+	Location operator+(const Location& other) const {
+		return { x + other.x, y + other.y };
+	}
+

```

Comment:
It is generally a good practice to separate the implementation from the header. Move the implementation of the Factory class member functions to a separate source file (cpp) and keep only the function declarations in the header file.